src/main/java/org/apache/mesos/scheduler/DefaultScheduler.java

```diff
@@ -0,0 +1,311 @@
+import org.apache.mesos.dcos.DcosConstants;
+import org.apache.mesos.offer.*;
+import org.apache.mesos.reconciliation.DefaultReconciler;
+import org.apache.mesos.reconciliation.Reconciler;
+import org.apache.mesos.scheduler.plan.*;
+import org.apache.mesos.scheduler.recovery.*;
+import org.apache.mesos.scheduler.recovery.constrain.LaunchConstrainer;
+import org.apache.mesos.scheduler.recovery.constrain.TimedLaunchConstrainer;
+import org.apache.mesos.scheduler.recovery.monitor.TimedFailureMonitor;
+import org.apache.mesos.specification.DefaultPlanSpecificationFactory;
+import org.apache.mesos.specification.PlanSpecification;
+import org.apache.mesos.specification.ServiceSpecification;
+import org.apache.mesos.state.PersistentOperationRecorder;
+import org.apache.mesos.state.StateStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.util.*;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * This scheduler when provided with a ServiceSpecification will deploy the service and recover from encountered faults
+ * when possible.  Changes to the ServiceSpecification will result in rolling configuration updates, or the creation of
+ * new Tasks where applicable.
+ */
+public class DefaultScheduler implements Scheduler {
+    private static final Integer DELAY_BETWEEN_DESTRUCTIVE_RECOVERIES_SEC = 10 * 60;
+    private static final Integer PERMANENT_FAILURE_DELAY_SEC = 20 * 60;
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private final ServiceSpecification serviceSpecification;
+    private final String zkConnectionString;
+    private final ExecutorService executor = Executors.newFixedThreadPool(1);
+    private final AtomicReference<RecoveryStatus> recoveryStatusRef;
+
+    private Reconciler reconciler;
+    private StateStore stateStore;
+    private TaskFailureListener taskFailureListener;
+    private TaskKiller taskKiller;
+    private OfferAccepter offerAccepter;
+    private Plan plan;
+    private PlanManager planManager;
+    private DefaultPlanScheduler planScheduler;
+    private DefaultRecoveryScheduler recoveryScheduler;
+
+    public DefaultScheduler(ServiceSpecification serviceSpecification) {
+        this(serviceSpecification, DcosConstants.MESOS_MASTER_ZK_CONNECTION_STRING);
+    }
+
+    public DefaultScheduler(ServiceSpecification serviceSpecification, String zkConnectionString) {
+        this.serviceSpecification = serviceSpecification;
+        this.zkConnectionString = zkConnectionString;
+        recoveryStatusRef = new AtomicReference<>(new RecoveryStatus(Collections.emptyList(), Collections.emptyList()));
+    }
+
+    void awaitTermination() throws InterruptedException {
+        executor.shutdown();
+        executor.awaitTermination(10, TimeUnit.SECONDS);
```

Comment:
I'll make it a constant. Configuration management is going to be a different PR.